lib/v8.js

```diff
@@ -23,6 +23,59 @@ const {
 const { copy } = internalBinding('buffer');
 const { objectToString } = require('internal/util');
 const { FastBuffer } = require('internal/buffer');
+const { toPathIfFileURL } = require('internal/url');
+const { validatePath } = require('internal/fs/utils');
+const { toNamespacedPath } = require('path');
+const {
+  createHeapDumpStream,
+  triggerHeapDump
+} = internalBinding('heap_utils');
+const { Readable } = require('stream');
+const { owner_symbol } = require('internal/async_hooks').symbols;
+const {
+  kUpdateTimer,
+  onStreamRead,
+} = require('internal/stream_base_commons');
+const kHandle = Symbol('kHandle');
+
+
+function writeHeapSnapshot(filename) {
+  if (filename !== undefined) {
+    filename = toPathIfFileURL(filename);
+    validatePath(filename);
+    filename = toNamespacedPath(filename);
+  }
+  return triggerHeapDump(filename);
+}
+
+class HeapSnapshotStream extends Readable {
+  constructor(handle) {
+    super({ autoDestroy: true });
+    this[kHandle] = handle;
+    handle[owner_symbol] = this;
+    handle.onread = onStreamRead;
+  }
+
+  _read() {
+    if (this[kHandle])
+      this[kHandle].readStart();
+  }
+
+  _destroy() {
+    // Release the references on the handle so that
+    // it can be garbage collected.
+    this[kHandle][owner_symbol] = undefined;
+    this[kHandle] = undefined;
+  }
+
+  [kUpdateTimer]() {
+    // Does nothing
+  }
+}
+
+function getHeapSnapshot() {
+  return new HeapSnapshotStream(createHeapDumpStream());
```

Comment:
How do we extend this if we want to provide an API to the user that returns the underlying string synchronously in the future? Maybe we can return an object that has a `toStream()` method instead so that we can also override `toString()` later? (we could just get rid of `heap_utils.createHeapDump()` with something that do a `JSON.parse` on top of the string if we have that)

Reply:
One could easily just add a new method to the underlying object. It wouldn't be too difficult to do in a separate PR

Reply:
@joyeecheung @jasnell One could also use `v8.getHeapSnapshot().read()`, if you want to get the full JSON snapshot synchronously.